test/config-luatest/vshard_upgrade_test.lua

```diff
@@ -0,0 +1,222 @@
+    g.instance_4 = server:new(fun.chain(opts, {alias = 'instance-004'}):tomap())
+    g.instance_5 = server:new(fun.chain(opts, {alias = 'instance-005'}):tomap())
+    g.instance_1:start({wait_until_ready = false})
+    g.instance_2:start({wait_until_ready = false})
+    g.instance_3:start({wait_until_ready = false})
+    g.instance_4:start({wait_until_ready = false})
+    g.instance_5:start({wait_until_ready = false})
+    g.instance_1:wait_until_ready()
+    g.instance_2:wait_until_ready()
+    g.instance_3:wait_until_ready()
+    g.instance_4:wait_until_ready()
+    g.instance_5:wait_until_ready()
+
+    local exec = [[
+        function put(v)
+            box.space.a:insert({v.id, v.bucket_id})
+            return true
+        end
+
+        function get(id)
+            return box.space.a:get(id)
+        end
+    ]]
+    g.instance_1:eval(exec)
+    g.instance_2:eval(exec)
+    g.instance_3:eval(exec)
+    g.instance_4:eval(exec)
+end)
+
+local function check_vshard(g, id_1, id_2)
+    local write_exec = string.format([[
+        vshard.router.call(1, 'write', 'put', {{id = %s, bucket_id = 1}})
+        vshard.router.call(2000, 'write', 'put', {{id = %s, bucket_id = 2000}})
+    ]], id_1, id_2)
+    g.instance_5:eval(write_exec)
+
+    local read_exec = string.format([[
+        local t = {}
+        t[1] = vshard.router.call(1, 'read', 'get', {%s})
+        t[2] = vshard.router.call(2000, 'read', 'get', {%s})
+        return t
+    ]], id_1, id_2)
+    t.helpers.retrying({timeout = 20}, function()
+        local ret = g.instance_5:eval(read_exec)
+        t.assert_equals(ret[1], {id_1, 1})
+        t.assert_equals(ret[2], {id_2, 2000})
+    end)
+end
+
+local function wait_for_names(instance_name, replicaset_name)
+    t.helpers.retrying({timeout = 20}, function()
+        local info = box.info
+        t.assert_equals(info.name, instance_name)
+        t.assert_equals(info.replicaset.name, replicaset_name)
+    end)
+end
+
+g.test_vshard_upgrade = function(g)
+    local version_exec = [[return box.space._schema:get{'version'}]]
+    local version = g.instance_1:eval(version_exec)
+    t.assert_equals(version, {'version', 2, 11, 1})
+
+    -- Check, that vshard works before schema upgrade.
+    check_vshard(g, 1, 2000)
+
+    -- Upgrade schema.
+    local upgrade_exec = [[
+        box.ctl.wait_rw()
+        box.schema.upgrade()
+    ]]
+    g.instance_1:eval(upgrade_exec)
+    g.instance_3:eval(upgrade_exec)
+
+    -- Check, that vshard works before all names are set.
+    check_vshard(g, 2, 2001)
+    -- Wait for names to be applied.
+    local rs_prefix = 'replicaset-00'
+    g.instance_1:exec(wait_for_names, {g.instance_1.alias, rs_prefix .. '1'})
+    g.instance_2:exec(wait_for_names, {g.instance_2.alias, rs_prefix .. '1'})
+    g.instance_3:exec(wait_for_names, {g.instance_3.alias, rs_prefix .. '2'})
+    g.instance_4:exec(wait_for_names, {g.instance_4.alias, rs_prefix .. '2'})
```

Comment:
> -- Check, that vshard works before all names are set.
> check_vshard(g, 2, 2001)

We can't say whether the check occurs before and after names are set. Especially due to retrying in the check.

Don't know, whether we should do anything in this regard.

Reply:
Yeah, this check is "flaky": it can occur before any names were applied, when some names were already applied, or even after all names were applied. We don't have errinjections in config module yet, and I didn't want to introduce one for the test.

But I want to be sure, that vshard works in all these cases, so I'd rather not remove it. The test is stable, so I suppose, it works in all cases

Reply:
That's a way to go. I would just clarify it in a comment in the code.